site-content/source/modules/ROOT/pages/development/ide.adoc

```diff
@@ -2,68 +2,84 @@
+
+=== Getting the source code
 
 The source code for Cassandra is shared on the central Apache Git
 repository and organized by branch, one branch for each major version.
-You can access the code for the current development branch using:
+You can access the code for the current development branch at
+https://gitbox.apache.org/repos/asf/cassandra.git (the primary location)
+or at https://github.com/apache/cassandra (a mirror location).
 
-// 01.15.21 - polandll - should this step be a fork rather than a clone?
+However, to develop a patch or a new feature, you should fork the Cassandra
+project first and then clone it from your own fork:
 
-[source, plaintext]
+[source,plaintext]
 ----
-git clone https://gitbox.apache.org/repos/asf/cassandra.git cassandra-trunk
+git clone https://github.com/<your_git_name>/cassandra.git cassandra
 ----
 
-Other branches will point to different versions of Cassandra. Switching
-to a different branch requires checking out the branch.
-For example, to checkout the latest version of Cassandra 3.0, use:
+=== Building Cassandra
 
-[source, plaintext]
-----
-git checkout cassandra-3.0
-----
+Cassandra is a Java project which is built using Ant.
+The build file, `build.xml` located in the root of the project content,
+has various tasks defined (you can list all of them with some short
+descriptions by running `ant -p`).
 
-You can get a list of available branches with `git branch`.
+The build uses the Java compiler which is currently set up in your
+shell. By default, the build uses Java 8. If you want to build
+with Java 11, you need to either add a build property `-Duse.jdk11=true`
+to your Ant build command or export the environment variable
+`CASSANDRA_USE_JDK11=true`. Otherwise, if you run the build using
+Java 11, the build script complains.
 
-Build Cassandra using ant:
+Now you can build Cassandra with the default task - just execute
+`ant` or `ant jar`. This may take a significant amount of time depending
+on artifacts that have to be downloaded or the number of classes that
+need to be compiled. When the build completes, you can find a JAR file
+in the build directory, and the database service can be started with
+the `bin/cassandra` script.
 
-[source, plaintext]
-----
-ant
-----
+=== Building a distribution
 
-This may take a significant amount of time depending on artifacts that have to
-be downloaded or the number of classes that need to be compiled.
+Some tasks you might be interested in are:
+
+* `sources-jar` - produce accompanying JAR for the primary Cassandra
+artifact (also in the build directory)
+* `artifacts` - generate Cassandra distribution in `build/dist`
+directory and package it to `tar.gz` (with and without sources)
 
 [TIP]
 .Hint
 ====
-You can setup multiple working trees for different Cassandra versions
-from the same repository using
-https://git-scm.com/docs/git-worktree[git-worktree].
+Remember to have only one Cassandra JAR artifact in the build directory;
```

Comment:
This is still a bit clumsy, because there are always multiple Cassandra JAR artefacts in the build directory, after any build.

e.g. something like… 
```
Remember to have only one versions of the different built jars in the build directory.
```

Reply:
No, when you just do build with default `jar` task, you have only one `apache-cassandra*.jar` in the `build` directory - so other versions, but also no sources or javadoc jars. This is extremely important because of this part of `casandra.in.sh`:

```shell
if [ -d $CASSANDRA_HOME/build ] ; then
    #cassandra_bin="$CASSANDRA_HOME/build/classes/main"
    cassandra_bin=`ls -1 $CASSANDRA_HOME/build/apache-cassandra*.jar`

    CLASSPATH="$CLASSPATH:$cassandra_bin"
fi
```

In particular, when `ls -1 $CASSANDRA_HOME/build/apache-cassandra*.jar` returns multiple results, they will be split by new line character and completely ruin the classpath. Perhaps that needs some care in Cassandra code, but such implementation is in existing Cassandra versions thus for the documentation purposes it is safe to say that we can only have a single `apache-cassandra*.jar` in the `build` directory. When there are more, you get `Error: Could not find or load main class org.apache.cassandra.service.CassandraDaemon` and you would  be probably like what the heck is wrong, because everything seems to be ok - project is properly built, tests are passing, all jars were generated...

Reply:
`ant artifacts` generates the following `apache-cassandra*.jar` files…
```
❯ ls build/apache-cassandra*.jar
build/apache-cassandra-4.0.2-SNAPSHOT-javadoc.jar build/apache-cassandra-4.0.2-SNAPSHOT-sources.jar build/apache-cassandra-4.0.2-SNAPSHOT.jar
```